Road name gap check

smaheshwaram opened the pull request:
### Description:

This check identifies Edges with no name Tag that are between two other Edges with the same name Tag, OR the Edge has a name Tag but does not equal the name Tag of the Edges that it is between.

i) This check should flag the edges who has NO name tag with below conditions:
   1. Is an Edge
   2. Has the Tag,
      a. highway=TERTIARY, SECONDARY, PRIMARY, TRUNK, or MOTORWAY
   3. Does NOT have the Tag,
      a. name=*
      b. highway=*_LINK
         1. TERTIARY_LINK, SECONDARY_LINK, PRIMARY_LINK, TRUNK_LINK, or MOTORWAY_LINK
      c. junction=ROUNDABOUT
   
ii) This check should flag the edges which has different Name Tag:
    1. Is an Edge
    2. Has the Tag,
        1. highway=TERTIARY, SECONDARY, PRIMARY, TRUNK, or MOTORWAY.
    3. name=* but is not the same as the other Edges it is connected to
    4. Does NOT have the Tag,
            a. highway=*_LINK
            b. junction=ROUNDABOUT

### Potential Impact:

A new check will be added to Atlas-checks project to identify and flag the edges with above mentioned description.

### Unit Test Approach:

Unit tests are added for the new check. Classes including RoadNameGapCheckTest.java and also a Ruler class i.e., RoadNameGapCheckTestRule to set up Atlas data.

### Test Results:

The code has been tested thoroughly to address the above mentioned description. Tested all the use cases. Tested all validations mentioned above in the description by taking 30MB data atlas files.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/atlas/predicates/TagPredicates.java:
```diff
@@ -53,6 +53,26 @@ private TagPredicates()
+     * Tests if the {@link AtlasObject} has a highway tag that do not contain TERTIARY_LINK,
+     * SECONDARY_LINK, PRIMARY_LINK, TRUNK_LINK, or MOTORWAY_LINK
+     */
+    public static final Predicate<AtlasObject> IS_HIGHWAY_NOT_LINK_TYPE = object -> Validators
```
`isLink()` method in HighwayTag can be leveraged to check if the highway type is a link type or not.

smaheshwaram replied:
Sure. I will take a look.

Bentleysb replied:
Is this being used anymore?

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/atlas/predicates/TagPredicates.java:
```diff
@@ -53,6 +53,26 @@ private TagPredicates()
+     * Tests if the {@link AtlasObject} has a highway tag that do contain TERTIARY, SECONDARY,
+     * PRIMARY, TRUNK, or MOTORWAY
+     */
+    public static final Predicate<AtlasObject> VALID_HIGHWAY_TAG = object -> Validators
```
I feel like this is specific to this check and should go within the check class than in the TagPredicates. Valid highways for this check may not be the same for other checks.

smaheshwaram replied:
Makes sense. Will move it to current file.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,134 @@
+{
+
+    // You can use serialver to regenerate the serial UID.
+    private static final long serialVersionUID = 1L;
```
This needs to be regenerated.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,134 @@
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+        final Set<Edge> inEdges = edge.inEdges().stream().filter(this::validCheckForObject)
```
Since the inEdges and outEdges sets are used only to collect name tags, we can filter and collect the inEdgesNameTags and outEdgesNameTags instead of inEdges and outEdges and simply compare the two sets to get set of matching name tags. This will simplify the findInEdgeOutEdgeMatchingName().

smaheshwaram replied:
Are you saying that there is a method which gets in inEdgeNameTags()? Please correct me if my understanding is wrong.

sayas01 replied:
What I meant is if we do something like this:
```final Set<String> inEdgesNameTags = edge.inEdges().stream().filter(inEdge -> this.validCheckForObject(inEdge) && NameTag.getNameOf(inEdge).isPresent()).map(inEdge -> NameTag.getNameOf(inEdge).get()).collect(Collectors.toSet());```
for both inEdges and outEdges, then we can simply compare the two sets to get matching name tags.

smaheshwaram replied:
Got it. I just thought for a second that there is a default method which I am not aware of. Thanks.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,134 @@
+        return Optional.empty();
+    }
+
+    private Set<String> findInEdgeOutEdgeMatchingName(final Set<Edge> inEdges,
```
Please see my previous comment on simplifying this method.

danielduhh commented on config/configuration.json:
```diff
@@ -444,6 +444,23 @@
     }
   },
+  "RoadNameGapCheck": {
+    "tag.filters":[
```
Are these tag filters being used?

smaheshwaram replied:
We are not using it. I thought its necessary to add this filters in json file. But, will take it out.

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/validation/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,134 @@
+            return Optional.empty();
+        }
+
+        final Set<String> edgeNames = findInEdgeOutEdgeMatchingName(inEdges, outEdges);
```
nit: Use `this.` when calling class methods

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/validation/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,134 @@
+        final Optional<String> edgeName = edge.getName();
+        if (edgeName.isPresent() && !edgeNames.contains(edgeName.get()))
+        {
+            final String instruction = "Edge name is different from in edge name and out edge name.";
```
Adding both the flagged Edge's name and the name of the mismatched in/out Edges will be useful for community mappers and debugging.

danielduhh replied:
`getLocalizedInstruction` is a useful function for handling instructions.

You can declare your instructions as a static global variables [like this](https://github.com/osmlab/atlas-checks/blob/dev/src/main/java/org/openstreetmap/atlas/checks/validation/tag/InvalidTagsCheck.java#L36)
and fetch the values [like this](https://github.com/osmlab/atlas-checks/blob/dev/src/main/java/org/openstreetmap/atlas/checks/validation/tag/InvalidTagsCheck.java#L145)

seancoulter commented on src/main/java/org/openstreetmap/atlas/checks/validation/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,134 @@
+        return object instanceof Edge && Edge.isMasterEdgeIdentifier(object.getIdentifier())
+                && TagPredicates.IS_HIGHWAY_NOT_LINK_TYPE.test(object)
+                && TagPredicates.VALID_HIGHWAY_TAG.test(object)
+                && HighwayTag.isCarNavigableHighway(object) && !JunctionTag.isRoundabout(object);
```
Since all tags checked in ```VALID_HIGHWAY_TAG``` are contained in the set of car navigable highways, we could just call ```TagPredicates.VALID_HIGHWAY_TAG.test(object)``` without ```HighwayTag.isCarNavigableHighway(object)```

seancoulter replied:
Also, would we want to check for the missing name tag here? For ```object```

smaheshwaram replied:
The requirement mentioned about all the valid tags along with car navigable highway so I was using the default method. Will take it out. Thanks.

seancoulter commented on src/test/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheckTestRule.java:
```diff
@@ -0,0 +1,116 @@
+
+    @TestAtlas(
+            // nodes
+            nodes = { @TestAtlas.Node(coordinates = @TestAtlas.Loc(value = TEST_1), id = "0"),
```
nit: we've already imported TestAtlas.Node, so the compiler should recognize just ```@Node```

smaheshwaram replied:
Makes sense.

sayas01 commented on docs/checks/RoadNameGapCheck.md:
```diff
@@ -0,0 +1,37 @@
+# Road name gap check
```
We also have a document https://github.com/osmlab/atlas-checks/blob/dev/docs/available_checks.md with all the available checks. Could you add add this check to the available_checks.md doc as well?

smaheshwaram replied:
Done.

smaheshwaram commented:
@danielduhh Dan got the analysis report and there are 2 scenarios which I need to take care of. All the requirements are met. However, there are 2 things which we didn't think of as part of requirement and it needs to be handled. I will work on it. Please don't merge it until then. Thanks.

danielduhh commented:
@smaheshwaram Sounds good. Please prepend [DNM] to your PR name  as an indicator that your check is still a work in progress. Thanks!

smaheshwaram commented:
Please look into this below link to understand the cases. Only valid cases are case 1 and case 4. Ignore case 2 and 3.

https://critigen.quip.com/edilAybbS3Dn/RoadNameGapCheck-Analysis

smaheshwaram commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,201 @@
+                    createFlag(object, this.getLocalizedInstruction(0, edge.getOsmIdentifier())));
+        }
+        final Optional<String> edgeName = edge.getName();
+
```
The code in this if condition is handling the 2 cases. Please take a look at getMatchingInAndOutEdgeNames() method and also finaMatchingEdgeNameWithAnyInEdgeOrOutEdge() method.

smaheshwaram commented:
@danielduhh Analysis on this check is done. Failing cases are resolved. Dan Castro please confirm it.

Bentleysb commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,218 @@
+     * PRIMARY, TRUNK, or MOTORWAY
+     */
+    private static final Predicate<AtlasObject> VALID_HIGHWAY_TAG = object -> Validators
+            .hasValuesFor(object, HighwayTag.class)
```
I think this is redundant when using Validators.isOfType().

smaheshwaram replied:
Okay. I will take this out.

Bentleysb commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,218 @@
+     */
+    private static final Predicate<AtlasObject> VALID_HIGHWAY_TAG = object -> Validators
+            .hasValuesFor(object, HighwayTag.class)
+            && Validators.isOfType(object, HighwayTag.class, HighwayTag.TERTIARY,
```
It might be good to make these values configurable, in case someone want to expand the range of this check at some point.

smaheshwaram replied:
Do you mean in the json file?

Bentleysb replied:
Yes

smaheshwaram replied:
Done

seancoulter commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,218 @@
+    /**
+     * This checks if the connected edges have same direction or not.
+     */
+    class EdgePredicate
```
Nice use of EdgeDirectionComparator. Why do we wrap EdgeDirectionComparator in EdgePredicate? Seems like we could just use the former directly

smaheshwaram replied:
I use EdgePredicate class for reusability purpose. I was using filter in flag() method in two places one for inEdge and one for outEdge. Do not like to repeat the code. Cannot add a method outside it because I need two parameters for comparison.

seancoulter commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,218 @@
+            return Optional.of(
+                    createFlag(object, this.getLocalizedInstruction(0, edge.getOsmIdentifier())));
+        }
+        final Optional<String> edgeName = edge.getName();
```
We can just get() the name directly here since we've verified it exists in the block above

smaheshwaram replied:
I can only use edge.getName().get() directly instead of using optional which is resulting in sonar issue. So, used this approach to handle it.

seancoulter replied:
It throws a sonar error if you do ```final String edgeName = edge.getName().get()``` on line 130?

smaheshwaram replied:
It says call Optional isPresent() before accessing the value. So, I coded this way.

seancoulter commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,218 @@
+
+        if (edgeName.isPresent() && !matchingInAndOutEdgeNames.contains(edgeName.get()))
+        {
+            // Case 1: Edge name: Tai. Incoming edge names: Tai , Shai. Outgoing edge name: Shai.
```
These are cases we want to flag, right? If so, we can probably indicate that and I'd vote to move to the docs

smaheshwaram replied:
Moved it to RoadNameGapCheck.md file case examples.

seancoulter commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,218 @@
+            // Pendler.
+            final Set<Edge> connectedEdges = edge.connectedEdges().stream()
+                    .filter(this::validCheckForObject).collect(Collectors.toSet());
+            if (findMatchingEdgeNameWithConnectedEdges(connectedEdges, edgeName.get()))
```
Here it looks like we don't flag edges if one of its in/out edges shares its name. If that's the case I think this ```if``` block can be moved to line 113 or so since it doesn't seem necessary to wait for getMatchingInAndOutEdgeNames() to finish.

smaheshwaram replied:
I want to do this validation only when there is at least one matching in and out edge name. Because, there was a failing case where there is matching in and out edge but the interconnecting edge is connected to another edge which has a different name and which doesn't have either incoming or outgoing edge for it.

sayas01 commented on docs/available_checks.md:
```diff
@@ -56,6 +56,7 @@ This document is a list of tables with a description and link to documentation f
 | [InvalidAccessTagCheck](checks/invalidAccessTagCheck.md) | The purpose of this check is to identify invalid access tags. |
 | [InvalidLanesTagCheck](docs/checks/invalidLanesTagCheck.md) | The purpose of this check is to identify highways in OSM with an invalid lanes tag value. |
 | [MixedCaseNameCheck](checks/mixedCaseNameCheck.md) | The purpose of this check is to identify names that contain invalid mixed cases so that they can be edited to be the standard format. |
+| [RoadNameGapCheck](checks/RoadNameGapCheck.md) | The purpose of this check is to identify edge connected between two edges whose name tag is same. Flag the edge if the edge has a name tag different to name tag of edges connected to it or if there is no name tag itself.
```
typo: The purpose of this check is to identify edge connected between two edges whose name tag is ~same~ different.

smaheshwaram replied:
Actually, it is not a typo. I need to find an edge which is connected between two edges whose name tag is same. IF the intermediate edge has no name then flag it. If the edge has a name but is different to the matching in and out edge then check if the edge has any other connected edge which has same name. If no then create a flag.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,210 @@
+        {
+            final Set<Edge> connectedEdges = edge.connectedEdges().stream()
+                    .filter(this::validCheckForObject).collect(Collectors.toSet());
+            if (findMatchingEdgeNameWithConnectedEdges(connectedEdges, edgeName.get()))
```
nit: Could be simplified using ternary operator.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,210 @@
+    private boolean findMatchingEdgeNameWithConnectedEdges(final Set<Edge> connectedEdges,
+            final String edgeName)
+    {
+        for (final Edge connectedEdge : connectedEdges)
```
Could make use of stream to replace this for loop. ```return connectedEdges.stream().anyMatch(....)```

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,210 @@
+        final Edge edge = (Edge) object;
+        final EdgePredicate edgePredicate = new EdgePredicate(edge);
+        final Set<Edge> inEdges = edge.inEdges().stream().filter(this::validCheckForObject)
+                .filter(edgePredicate::isSameHeading).collect(Collectors.toSet());
```
Couldn't we just use a single filter and put validCheckForObject and isSameHeading in the same filter?

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,210 @@
+        final Set<String> edgeNames = new HashSet<>();
+        for (final Edge inEdge : inEdges)
+        {
+            if (!inEdge.getName().isPresent())
```
It would be cleaner if we could add this condition in the line that collects the inEdge and outEdge. 
```edge.inEdges().stream().filter(edge -> edge.getName().isPresent() && this.validCheckForObject(edge) &&....)```

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,210 @@
+            }
+            for (final Edge outEdge : outEdges)
+            {
+                if (!outEdge.getName().isPresent())
```
Same comment as above.

sayas01 commented on src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java:
```diff
@@ -0,0 +1,210 @@
+    private static final long serialVersionUID = 7104778218412127847L;
+    private static final List<String> VALID_HIGHWAY_TAG_DEFAULT = Arrays.asList("primary",
+            "secondary", "tertiary", "trunk", "motorway");
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays.asList("Edge name is empty.",
```
nit: It would be useful to have the flagged object's identifier in the instruction.

smaheshwaram commented:
![Screen Shot 2020-01-07 at 4 41 28 PM](https://user-images.githubusercontent.com/16053580/71935589-cafc9000-316c-11ea-9419-04498ccf530e.png)

@danielduhh Code coverage is showing that it covered 48 lines out of 51. However, sonar quality gates failing on 0% code coverage. Build was fine until today morning. I was just addressing PR comments. Not major changes in code. So, please check if something needs to be fixed from my end.

danielduhh commented:
Hey! I'm not sure why Sonar is throwing fits.. the unit tests are clearly there @smaheshwaram . We can try opening and closing the PR to retrigger the build

danielduhh commented:
@mgcuthbert do you have any idea what's going on with Sonar? It's reporting 0% code coverage...

yiqingj commented:
it does look like a sonar issue always reporting 0% coverage. I'll merge this one manually for now to unblock the release.
